File: app/src/androidTest/java/com/sotwtm/util/ECLogcatUtilTest.java
package com.sotwtm.util;

import android.content.Context;
import android.os.Environment;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.text.TextUtils;

import junit.framework.TestCase;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;

/**
 * Test case for {@link ECLogcatUtil}
 * @author sheungon
 */
@RunWith(AndroidJUnit4.class)
public class ECLogcatUtilTest extends TestCase {

    static final String TAG = "ECLogcatUtilTest";

    @Before
    public void clearSettings() throws Exception {
        Context context = InstrumentationRegistry.getTargetContext();
        context.getSharedPreferences(ECLogcatUtil.SHARED_PREF_FILE_KEY, Context.MODE_PRIVATE).edit().clear().apply();
    }

    @Test
    public void testLogcat() throws Exception {

        Context context = InstrumentationRegistry.getTargetContext();
        ECLogcatUtil logcatUtil = ECLogcatUtil.getInstance(context);

        assertNotNull(logcatUtil);

        try {
            logcatUtil.resetLogcat();
            assertTrue(false);
        } catch (NullPointerException e) {
            assertTrue(true);
        }

        try {
            logcatUtil.startLogcat();
            assertTrue(false);
        } catch (NullPointerException e) {
            assertTrue(true);
        }

        try {
            logcatUtil.clearLogcat();
            assertTrue(false);
        } catch (NullPointerException e) {
            assertTrue(true);
        }

        String appRunByUser = ECLogcatUtil.getAppRunByUser(context);
        assertFalse(TextUtils.isEmpty(appRunByUser));
        assertNull(ECLogcatUtil.getLogcatPIDRunningBy(appRunByUser));
        assertFalse(ECLogcatUtil.isLogcatRunningBy(appRunByUser));

        File logFile = new File(Environment.getExternalStorageDirectory().getPath(), "log.txt");
        assertTrue(!logFile.isFile() || logFile.delete());
        logcatUtil.setLogcatDest(logFile);
        logcatUtil.setFilterLogTag(TAG);

        assertTrue(logcatUtil.startLogcat());
        Log.d(TAG, "This is a test log.");
        Thread.sleep(100);

        assertNotNull(ECLogcatUtil.getLogcatPIDRunningBy(appRunByUser));
        assertTrue(ECLogcatUtil.isLogcatRunningBy(appRunByUser));

        assertTrue(logFile.isFile());
        assertTrue(logcatUtil.stopLogcat());
        assertFalse(ECLogcatUtil.isLogcatRunningBy(appRunByUser));
    }
}

File: ec-log/src/androidTest/java/com/sotwtm/util/ECLogcatUtilTest.kt
package com.sotwtm.util

import android.app.Application
import android.content.Context
import android.support.test.InstrumentationRegistry
import android.support.test.runner.AndroidJUnit4
import junit.framework.TestCase
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.File

/**
 * Test case for [ECLogcatUtil]
 * @author sheungon
 */
@RunWith(AndroidJUnit4::class)
class ECLogcatUtilTest : TestCase() {

    /**
     * Clean up SharedPreferences before test
     * */
    @Before
    @Throws(Exception::class)
    fun clearSettings() {
        val context = InstrumentationRegistry.getTargetContext()
        context.getSharedPreferences(ECLogcatUtil.SHARED_PREF_FILE_KEY, Context.MODE_PRIVATE).edit().clear().apply()
    }

    /**
     * Test case for [ECLogcatUtil] to check if it works normally.
     * */
    @Test
    @Throws(Exception::class)
    fun testLogcat() {

        val application = InstrumentationRegistry.getTargetContext().applicationContext as Application
        val logcatUtil = ECLogcatUtil.getInstance(application)

        Assert.assertNotNull(logcatUtil)

        try {
            logcatUtil.resetLogcat()
            Assert.assertTrue(false)
        } catch (e: NullPointerException) {
            Assert.assertTrue(true)
        }

        try {
            logcatUtil.startLogcat()
            Assert.assertTrue(false)
        } catch (e: NullPointerException) {
            Assert.assertTrue(true)
        }

        try {
            logcatUtil.clearLogcat()
            Assert.assertTrue(false)
        } catch (e: NullPointerException) {
            Assert.assertTrue(true)
        }

        val appRunByUser = ECLogcatUtil.getAppRunByUser(application)
        if (appRunByUser == null) {
            Assert.assertNotNull(appRunByUser)
            return
        }
        Assert.assertFalse(appRunByUser.isEmpty())
        Assert.assertNull(ECLogcatUtil.getLogcatPIDRunBy(appRunByUser))
        Assert.assertFalse(ECLogcatUtil.isLogcatRunningBy(appRunByUser))

        val logFile = File(application.cacheDir, "log.txt")
        Assert.assertTrue(!logFile.isFile || logFile.delete())
        logcatUtil.setLogcatDest(logFile)
        logcatUtil.setFilterLogTag(TAG)

        Assert.assertTrue(logcatUtil.startLogcat())
        Log.d(TAG, "This is a test log.")
        Thread.sleep(100)

        Assert.assertNotNull(ECLogcatUtil.getLogcatPIDRunBy(appRunByUser))
        Assert.assertTrue(ECLogcatUtil.isLogcatRunningBy(appRunByUser))

        Assert.assertTrue(logFile.isFile)
        Assert.assertTrue(logcatUtil.stopLogcat())
        Assert.assertFalse(ECLogcatUtil.isLogcatRunningBy(appRunByUser))
    }

    companion object {
        internal const val TAG = "ECLogcatUtilTest"
    }
}